### PR TITLE
refactor(ndt_scan_matcher): isolate align function into ndt_omp

### PR DESCRIPTION
**Pull request by kminoda**
Signed-off-by: kminoda <koji.minoda@tier4.jp>

## Description
I would like to isolate the align function in the current ndt_scan_matcher (and also the `NdtResult` class) to `ndt_omp`.

This PR should be merged at the same time as https://github.com/tier4/ndt_omp/pull/16.

(side note: This PR may induce a build error due to the version mismatch between autoware.universe and ndt_omp, only when the newer universe used with the older ndt_omp)

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].
- [ ] The PR has been properly tested.
- [ ] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

**Comment by kminoda** [comment]
This PR will be ready for review after https://github.com/tier4/ndt_omp/pull/16 is merged. (The CI test would not pass until then)

**Review by KYabuuchi — COMMENTED** [review]
Thanks for your contribution!
I left some comments. Please check them out.

File: localization/ndt_scan_matcher/src/ndt_scan_matcher_core.cpp [hunk 1]
```diff
@@ -429,14 +429,23 @@ void NDTScanMatcher::callback_sensor_points(
...
   key_value_stdmap_["state"] = "Sleeping";
 
   const auto exe_end_time = std::chrono::system_clock::now();
   const double exe_time =
     std::chrono::duration_cast<std::chrono::microseconds>(exe_end_time - exe_start_time).count() /
     1000.0;
 
+  const geometry_msgs::msg::Pose result_pose_msg = matrix4f_to_pose(ndt_result.pose);
+  std::vector<geometry_msgs::msg::Pose> transformation_msg_array;
+  for (auto pose_matrix : ndt_result.transformation_array) {
```
Comment (KYabuuchi) [inline localization/ndt_scan_matcher/src/ndt_scan_matcher_core.cpp:444]: I think `for (const auto& pose_matrix : ndt_result.transformation_array)` is better.

File: localization/ndt_scan_matcher/src/ndt_scan_matcher_core.cpp [hunk 2]
```diff
@@ -429,14 +429,23 @@ void NDTScanMatcher::callback_sensor_points(
 
   // perform ndt scan matching
   key_value_stdmap_["state"] = "Aligning";
-  const NdtResult ndt_result = align(interpolator.get_current_pose().pose.pose);
+  const Eigen::Matrix4f initial_pose_matrix =
+    pose_to_matrix4f(interpolator.get_current_pose().pose.pose);
+  const pclomp::NdtResult ndt_result = ndt_ptr_->executeScanMatching(initial_pose_matrix);
```
[the remark this answers is not in the document]
Comment (kminoda) [reply localization/ndt_scan_matcher/src/ndt_scan_matcher_core.cpp]: https://github.com/autowarefoundation/autoware.universe/pull/2199/commits/b844ccf0ad9a39ca0b294f28ebf49a68fb13f98b
Resolved

File: localization/ndt_scan_matcher/src/ndt_scan_matcher_core.cpp (hunk 1 above)
Comment (kminoda) [reply localization/ndt_scan_matcher/src/ndt_scan_matcher_core.cpp]: https://github.com/autowarefoundation/autoware.universe/pull/2199/commits/717c6161d41f4f0c6f123fbe27849a14d01344f9
Resolved

File: localization/ndt_scan_matcher/src/ndt_scan_matcher_core.cpp [hunk 3]
```diff
@@ -559,18 +545,21 @@ geometry_msgs::msg::PoseWithCovarianceStamped NDTScanMatcher::align_using_monte_
 
   for (unsigned int i = 0; i < initial_poses.size(); i++) {
     const auto & initial_pose = initial_poses[i];
-    const NdtResult ndt_result = align(initial_pose);
+    const Eigen::Matrix4f initial_pose_matrix = pose_to_matrix4f(initial_pose);
+    auto output_cloud = std::make_shared<pcl::PointCloud<PointSource>>();
```
Comment (KYabuuchi) [inline localization/ndt_scan_matcher/src/ndt_scan_matcher_core.cpp:549]: `output_cloud` is defined in line 544, too.  
So, please remove the `output_cloud` defined in the for-loop.

**Review by KYabuuchi — COMMENTED** [review]
Your PR is mostly fine :+1: ,  but I found one small mistake. Please correct it.
When it is resolved I will approve this PR soon.

File: localization/ndt_scan_matcher/src/ndt_scan_matcher_core.cpp (hunk 3 above)
Comment (kminoda) [reply localization/ndt_scan_matcher/src/ndt_scan_matcher_core.cpp]: Thanks! Resolved here: https://github.com/autowarefoundation/autoware.universe/pull/2199/commits/6580214b98dbe68c20c7347b1fe5434cba74fc09

**Review by KYabuuchi — APPROVED** [review]

**Pull request merged**